Keep WebGLExtension::m_context on stack to handle context loss.

djg opened the pull request:
#### 554972c0f0a7bbc17463fb2a024414d4ab0a09ad
<pre>
Keep WebGLExtension::m_context on stack to handle context loss.
<a href="https://bugs.webkit.org/show_bug.cgi?id=242788">https://bugs.webkit.org/show_bug.cgi?id=242788</a>
rdar://95969241

Reviewed by Kenneth Russell and Dean Jackson.

WebGLExtension::loseParentContext() nulls m_context pointer. In the GPUP world,
where any GL call can result in a context loss, this behaviour can lead to null
pointer access at random places inside a WebGL extension object after the code
has checked for a valid m_context pointer at function entry.

This change adds a stack-based RAII helper WebGLExtensionScopedContext to keep a
reference to the calling context in the event of context loss nulling
WebGLExtension::m_context. Checking for context lost has been unified behind
this helper class.

* Source/WebCore/html/canvas/ANGLEInstancedArrays.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/OESDrawBuffersIndexed.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/OESVertexArrayObject.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLDebugShaders.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLDrawBuffers.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLDrawInstancedBaseVertexBaseInstance.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLExtension.cpp:
Implement WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLExtension.h:
(WebCore::WebGLExtension::isLostContext): Renaming of isLost to catch any
locations using WebGLExtension directly instead of WebGLExtensionScopedContext
(WebCore::WebGLExtension::m_context): Make member private to force access via
public API context() or WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLLoseContext.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLMultiDraw.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLMultiDrawInstancedBaseVertexBaseInstance.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
Rename isLost() -&gt; isLostContext()

Canonical link: <a href="https://commits.webkit.org/252810@main">https://commits.webkit.org/252810@main</a>
</pre>

kenrussell reviewed (approved):
Looks good to me.

FWIW Chromium's solution (yours is less code): https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/webgl/webgl_extension.h;l=39?q=WebGLExtensionScopedContext

grorg reviewed (approved):
r=me assuming test failures go away

djg committed:
Keep WebGLExtension::m_context on stack to handle context loss.
https://bugs.webkit.org/show_bug.cgi?id=242788
rdar://95969241

Reviewed by Kenneth Russell and Dean Jackson.

WebGLExtension::loseParentContext() nulls m_context pointer. In the GPUP world,
where any GL call can result in a context loss, this behaviour can lead to null
pointer access at random places inside a WebGL extension object after the code
has checked for a valid m_context pointer at function entry.

This change adds a stack-based RAII helper WebGLExtensionScopedContext to keep a
reference to the calling context in the event of context loss nulling
WebGLExtension::m_context. Checking for context lost has been unified behind
this helper class.

* Source/WebCore/html/canvas/ANGLEInstancedArrays.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/OESDrawBuffersIndexed.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/OESVertexArrayObject.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLDebugShaders.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLDrawBuffers.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLDrawInstancedBaseVertexBaseInstance.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLExtension.cpp:
Implement WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLExtension.h:
(WebCore::WebGLExtension::isLostContext): Renaming of isLost to catch any
locations using WebGLExtension directly instead of WebGLExtensionScopedContext
(WebCore::WebGLExtension::m_context): Make member private to force access via
public API context() or WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLLoseContext.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLMultiDraw.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLMultiDrawInstancedBaseVertexBaseInstance.cpp:
Use WebGLExtensionScopedContext
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
Rename isLost() -> isLostContext()

Canonical link: https://commits.webkit.org/252810@main
